### ISSUE TITLE
Pyranometer or general HASS metrics integration

### ISSUE THREAD

**Issue by chron0** [mj-xmr/SolOptXMR#55]
Hi... <3 this project - using miners as a diversion load has been on my roadmap for a long time. I havent tested it out yet but I sure will in the future. In  my setup excess PV will charge the battery, then the warm water tank and then any access should be directed to mining. This would be a full off-grid scenario, so the miners should immediately throttle down, when the - by nature - intermittent PV production goes down to to clouds and quickly ramp up again to not drain the battery. This would require to involve metrics from battery as well as a separate pyranometer readout so that the system can ramp up again, as the pv inverters throttle their output when not enough demand is changing the frequency of the island grid.

How has your experience in practice been with on/off mining on p2pool, iirc it is PPLNS which is not really kind to on/off mining?

If you guys would like to discuss more p2p, you can always join us in #monero-pools on libera :)

**Comment by mj-xmr** [comment]
Hi and thanks for your insights.

Yes, various measurements are planned as the 3rd milestone, which after some manual verification, will be automated as the 4th milestone.

I'm actually just starting with p2pool. Do you think that the PPLNS problem could be minimized by having just one low powered miner on all the time?

I'll gladly join the #monero-pools. Thanks for invitation :)

**Comment by mj-xmr** [comment]
Also, slowly warming up water in an isolated tank does sound like a very cheap and useful energy storage.

**Comment by chron0** [comment]
Yeah it's like a battery in the end as well. Either by a PWM driven old style resistive heating element or the "more modern" approach. by feeding an air-water heatpump and using the COP leverage of 2-4, so that 1kw electric will become 2-4kW thermal energy that is getting fed in.

As for PPLNS, in the end, I dont think it really matters, we would have to run some numbers against a real world PPLNS time window an see how much loss there is. Ideally, for our use-case PPS would be the optimal algo, getting paid for shares we deliver, when we deliver them. Also, even tho xmr is my favorite crypto implementation, especially with the p2pool considerations - I think it would make sense to be a little more agnostic with this and have the option to control other algos/miners via API as well. Like GPU rigs for other coins for instance.

**Comment by mj-xmr** [comment]
> or the "more modern" approach, by feeding an air-water heatpump and using the COP leverage of 2-4, so that 1kw electric will become 2-4kW thermal energy that is getting fed in.

Yep. That's the expertise, that the market needs currently, even if (in the best case) they tell us that "it's still to early".

> Also, even tho xmr is my favorite crypto implementation, especially with the p2pool considerations - I think it would make sense to be a little more agnostic with this and have the option to control other algos/miners via API as well. Like GPU rigs for other coins for instance.

Definitely yes, and it's possible.

However this is kind of a delicate subject, at least for now. The project is being funded by the Monero Community and I have to balance between satisfying various parties, that are involved here. The dimension, that you're mentioning, describes an individual smart miner, who isn't afraid of switching between coins to maximize the $ profit.

Please note, however, that another dimension to keep satisfied with the project is the "_Monero Collective_", which consists not only of the Miners, but also (and probably: mainly) of regular users, who just need to have their transactions confirmed securely. They care mostly not about the miners' individual profits, but about the security of the Monero network itself. What binds us together here are the troughs in the network's difficulty across time, that my project aims to keep as shallow as possible, thus lessening the risks of a 51% attack. After all, just in May, there was a swing of the net. diff. of ~23% from the peak to the trough. Here's some recent history to help visualizing the problem better, generated [by SolOptXMR](https://github.com/mj-xmr/SolOptXMR#network-difficulty):

![Network difficulty](https://user-images.githubusercontent.com/63722585/169827786-7dc548c1-6b46-49a9-a7ac-ca20605f1046.png)

The troughs are already being taken into account by the system, yielding a bonus or a penalty towards the collected hashes across the stated horizon, while still having the battery drainage prevention as the main goal. Obviously scooping such troughs also give the individual miner a measurable advantage.

So long story short, what I mean to say is, that it's possible to extend the project in this direction in future, and honestly, if we don't do it, somebody will fork the project and will do it anyway, so we really have no choice. However, at least for as long as the milestones paid upfront by the Monero Community aren't yet reached, I can't really imagine how I could deprive the Community of the network security advantage, that they deserve for their involvement. Later on, I will first ask the Community if they see it feasible to extend the project, like you described, mentioning the forking dilemma. 

All in all, I'm so happy, that you're the next person, who understands the nuances of the project.

**Comment by chron0** [comment]
Ah, sorry for the delay, I got busy with other off-grid foo. 

I didn't realize that it was funded thru FFS, as I've stopped following what goes on there since in most cases it seems like just a grabbing bazaar without much return, so I am rather happy that you seem to be an exception there to actually produce something of real value AND put monero first.

Given that monero is already at tail emission and (human) miners (as sad as this reality is) are mostly in for profit the amount of actual users will likely be low. In the end the majority of monero hashrate is likely still in the hands of botnets. One can think or judge that however one pleases but it's undeniably a "feature", when it comes to risk managing the networks ability to survive and keep transactions going. 

All I'm really saying would be: Great to keep the focus on monero, I fully support that, but it might pay in the long run to do this as modular and pluggable as possible to make future additions flexible and easy :)

**Comment by mj-xmr** [comment]
> I didn't realize that it was funded thru CCS, as I've stopped following what goes on there since in most cases it seems like just a grabbing bazaar without much return, so I am rather happy that you seem to be an exception there to actually produce something of real value AND put monero first.

Thank you very much!

> Given that monero is already at tail emission and (human) miners (as sad as this reality is) are mostly in for profit the amount of actual users will likely be low. In the end the majority of monero hashrate is likely still in the hands of botnets. One can think or judge that however one pleases but it's undeniably a "feature", when it comes to risk managing the networks ability to survive and keep transactions going.

Even if it's a niche, then if it hasn't been explored yet, I'm going in. Besides, during the impending blackouts, the solar farm owners will be the only ones who will be able to mine. I tend to think a bit ahead of time. The current situation is rarely interesting for me.

> All I'm really saying would be: Great to keep the focus on monero, I fully support that, but it might pay in the long run to do this as modular and pluggable as possible to make future additions flexible and easy :)

I write my code in a modular way, so even if a given area appears tightly coupled at a given time, be sure, that it's on my `watch list` ;)

**Comment by chron0** [comment]
> > Given that monero is already at tail emission and (human) miners (as sad as this reality is) are mostly in for profit the amount of actual users will likely be low. In the end the majority of monero hashrate is likely still in the hands of botnets. One can think or judge that however one pleases but it's undeniably a "feature", when it comes to risk managing the networks ability to survive and keep transactions going.

> Even if it's a niche, then if it hasn't been explored yet, I'm going in. Besides, during the impending blackouts, the solar farm owners will be the only ones who will be able to mine. I tend to think a bit ahead of time. The current situation is rarely interesting for me.

That depends on the system really. The majority of PV installations out there will NOT produce anything if the grid goes down since they are grid tied only. And all inverters must shut down if they dont get frequency from the grid to sync to. It's basically a grid demanded "safety feature" so that there is no backfeed into the grid in case of an outage. Many people are not aware of this and setting up a PV system to be actually off-grid capable takes extra installation steps and hardware as well as island capable inverters.
 
> > All I'm really saying would be: Great to keep the focus on monero, I fully support that, but it might pay in the long run to do this as modular and pluggable as possible to make future additions flexible and easy :)
> 
> I write my code in a modular way, so even if a given area appears tightly coupled at a given time, be sure, that it's on my _watch list_ ;)

<3 :)

**Comment by mj-xmr** [comment]
I get ya.

Well, I built myself an off-grid system, and would expect anybody to do the same in the end, but I have to admit, that the island-capable inverter was the most expensive piece of hardware here.
